### PR TITLE
core: Rewrite builder class signatures to avoid internal class

### PR DESCRIPTION
**Pull request by ejona86**
This provides us a path forward with #7211 (hiding
AbstractManagedChannelImplBuilder and AbstractServerImplBuilder) while
providing users a migration path to manage the ABI breakage (#7552). We
do a .class hack so that recompiling avoids the internal class reference
yet the old methods are still available.

Leaving the classes as-is causes javac to compile two versions of each
method, one returning the public class (e.g. ServerBuilder) and one
returning the internal class (e.g., AbstractServerImplBuilder). However,
we rewrite the signature that is used at compile time so that new
compilations will not reference internal-returning methods.

This is intended to be temporary, just to give a migration path. Once we
have given users some time to recompile we will remove this rewriting
and change the generics to use public classes.

-----

There's probably a way to make the guava version to be centrally defined in the settings.gradle. But it doesn't need to match the version we use elsewhere and we'll just be deleting it in the future. So it seems best to KISS and just have the version inline.

CC @njhill

### CODE REVIEW

**Comment by njhill** [comment]
nice! My vote would be to leave in until 2.x :)

File: core/build.gradle [hunk 1]
```diff
@@ -53,7 +63,47 @@ animalsniffer {
...
+    int i = Bytes.indexOf(haystack, needle);
+    assert i != -1;
+    byte[] result = new byte[haystack.length - needle.length + replacement.length];
+    System.arraycopy(haystack, 0, result, 0, i);
+    System.arraycopy(replacement, 0, result, i, replacement.length);
+    System.arraycopy(haystack, i + needle.length, result, i + replacement.length, haystack.length - i - needle.length);
+    return result;
+}
+
+def bigEndian(int value) {
```
Comment (sergiitk) [inline core/build.gradle:76]: Maybe something like
```suggestion
def bigEndianBytesShort(int value) {
```

To indicate this only works with 2-byte ints?

File: core/build.gradle [hunk 2]
```diff
@@ -53,7 +63,47 @@ animalsniffer {
...
+    // CONSTANT_Utf8_info. https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.4.7
+    byte[] needleBytes = Bytes.concat(
+        [1] as byte[], bigEndian(needle.length()), needle.getBytes(US_ASCII));
+    byte[] replacementBytes = Bytes.concat(
+        [1] as byte[], bigEndian(replacement.length()), replacement.getBytes(US_ASCII));
+    file.setBytes(replaceBytes(file.getBytes(), needleBytes, replacementBytes));
+}
+
 plugins.withId("java") {
+    compileJava {
```
Comment (sergiitk) [inline core/build.gradle:90]: Would it make sense to add a comment explaining the reason why we do this? Or, perhaps, a tracking ticket to capture the context, and the plan detailing when we're planning to remove this.

**Review by sergiitk — APPROVED** [review]

File: core/build.gradle (hunk 1 above)
Comment (ejona86) [reply core/build.gradle]: I went ahead and used the name `bigEndianShortBytes()`

**Pull request merged**